proposals/0010-vk-buffer-ref.md

```diff
@@ -0,0 +1,270 @@
+* Make it possible through SPIR-V reflection to determine which members of a struct accessed by BufferDeviceAddress are statically referenced and at what offset.  This is already possible for other data like cbuffers in order for shader tooling to be able to identify which elements are used and where to put them.
+
+## Proposed solution
+
+Our solution is to add a new builtin type in the vk namespace that is a pointer to a buffer of a given type, `vk::buffer_ptr<T,A>`. The template argument `T` must be a struct. `A` must be an integer and is the alignment in bytes of the pointer. If `A` is not specified, the alignment is assumed to be 16 bytes.
+
+This new type will have the following operations
+
+* Copy assignment and copy construction - These copy the value of the pointer from one variable to another.
+* Accessing members of T - This is done using the `.` operator.
+* Conversion to uint64_t - Return the integer representation of the address it points to. This can be used for null pointer checks.
+
+Note the operations that are not allowed:
+
+* There is no default construction. Every vk::buffer_ptr<T> is either contained in a global resource (like a cbuffer, ubo, or ssbo), or it must be constructed using the copy constructor. 
+* There is no equivalent to the `*` operator that will dereference the entire struct.
+* There is no conversion from uint64_t to vk:buffer_ptr.
+* There is no explicit pointer arithmetic. All addressing is implicitly done using the `.` pointer, or indexing an array in the struct T.
+* The comparison operators == and != are not supported for buffer pointers. Comparison to null pointer should be done by first converting to uint64_t.
+
+Most of these restrictions are there for safety. They minimize the possibility of getting an invalid pointer. If the `.` operator is used on a null pointer, the behaviour is undefined.
+
+When used as a member in a buffer, vk::buffer_ptr can be used to pass physical buffer addresses into a shader, and address and access buffer space with logical addressing, which allows tools such as spirv-opt, spirv-reflect and renderdoc to be able to better work with these shaders.
+
+For example, here is a shader using vk::buffer_ptr to do the same thing as the shader above using vk::RawBufferLoad. Note the natural, logical syntax of the reference:
+
+```c++
+
+struct Globals_s
+{
+      float4 g_vSomeConstantA;
+      float4 g_vTestFloat4;
+      float4 g_vSomeConstantB;
+};
+
+typedef vk::buffer_ptr<Globals_s> Globals_p;
+
+struct TestPushConstant_t
+{
+      Globals_p m_nBufferDeviceAddress;
+};
+
+[[vk::push_constant]] TestPushConstant_t g_PushConstants;
+
+float4 MainPs(void) : SV_Target0
+{
+      float4 vTest = g_PushConstants.m_nBufferDeviceAddress.g_vTestFloat4;
```

Comment:
The accessing syntax is irregular here. A `vk::buffer_ptr<T>` can’t be treated as member accessing `T`. I think there are two options:

(1) Add a `Get` method to `vk::buffer_ptr` which returns a reference. This would be consistent with some of the existing buffer accessors, and provides regular syntax which can be supported for all HLSL versions.
(2) For HLSL 202x a dereference (`->`) operator could be added to indirect to the members.

Reply:
> For HLSL 202x a dereference (->) operator could be added to indirect to the members.

Does that mean that "->" cannot be added today? If not, what approximate calendar could this be added and released?

Reply:
Regarding get(), would the above line then be?:

`float4 vTest = g_PushConstants.m_nBufferDeviceAddress.get().g_vTestFloat4;

Reply:
The support for dereference operators is tied to adding references, which hasn’t yet been done. HLSL 202x doesn’t have a schedule yet, but we’re already starting to do some preliminary work on it.

Yes, that is how the `Get()` method should work.

Reply:
Is it required to have the full reference support to use `->`?
For now parser seems to ignore MemberExpr->isArrow() and allow `.` and `->` (isArrow is just always false), and then fails in 
`tools/clang/lib/Sema/SemaExprCXX.cpp`

Could we check the type here and allow arrow operators for this type without having the full reference implem?
This this feature can being used with the proper syntax, and then real reference support can roll-in without disturbances?

Reply:
I like the idea of a get method. I have mentioned that I do not like the idea that we cannot access the entire struct.

Reply:
We don’t need to have references fully implemented to add `->` and `*` dereference operators to built-in types, but those operators will be HLSL 202x features, not available in the older language versions.

There are some other changes that are planned for 202x which will require significant reworking of overload resolution (see: https://github.com/microsoft/hlsl-specs/pull/34), and I think that we should expect that adding new operator overloads will depend on that.

Reply:
@llvm-beanz Can you please give an example of an existing HLSL method similar to your intent for get() here?

Reply:
The `RW` resource type subscript operators return references in the AST, which is roughly what you need to do here. The buffer `Load` methods also have some similarities.

Reply:
I am ok with a dereferencing method, but I have concerns about its spelling.

get() seems to imply a physical load of the entire buffer, which none of ".", "->" or "[]" do, at least in my mind.

For example, the following seems to imply that we would load the entire buffer, then select the member g_vTestFloat4:

`float4 vTest = g_PushConstants.m_nBufferDeviceAddress.get().g_vTestFloat4;`

Is there a better spelling we could come up with? One idea that is closer to what I would like is deref(). Would this be acceptable? Any other ideas?

To be clear, I don't have a problem with supporting a load of the entire buffer as @s-perron would like, but I don't want to imply it if we are not doing it, as in the case above.

Reply:
> We don’t need to have references fully implemented to add `->` and `*` dereference operators to built-in types, but those operators will be HLSL 202x features, not available in the older language versions.

Right, just unsure how the release of the HLSL 202x relates to this. Since we add a new type to the language, how would that be different from adding allowing the `->` operator for this type? Is it because you don't want to have the operator being tied to a specific type for now instead of a more generic semantic?

Agree with Greg, as the `get()` does seem to imply we load the full struct, and then hope for the optimizer to only load 1 field.

Reply:
OK. I am fine with get().

Reply:
Thanks for the thorough explanation!
for `get()`, missed the parallel with shared pointers. Good argument, fine with `get()`.

Thanks for the clarification on the frontier between HLSL version and shader model version. Those arguments make sense 😊